Fix for #4099 (geotagging not working, location doesn't display on front-end)

griffbrad opened the pull request:
I'm creating this pull request to get some early feedback on my work to address issue #4099.  Basically, Calypso was previously only saving geo_latitude and geo_longitude meta fields.  The equivalent geo-tagging feature from the wp-admin UI on WordPress.com, however, also includes geo_address and geo_public meta fields.  Without these, geo-tagging information is not actually displayed on the frontend (as reported in #4099).

To address this inconsistency, I've added a "Display Options" popover beneath the current location pane on the Calypso post editor sidebar.  If no location is set, this is completely hidden, leaving the UI exactly as-is for folks not actively using the location feature.  If you set a location, using either the geolocation API or the Google Maps search, a "Display Options" button is displayed beneath the current UI, allowing you to directly edit geo_public and geo_address.

geo_address is the human-friendly version (e.g. “Chicago, IL”) of the location that is actually displayed on the frontend rather than the coordinates (e.g. “41.878114, -87.629798”).  One known gap in my current approach in this feature branch is that geo_address does just get set to those raw coordinates when using the browser's geolocation API rather than performing a Google Maps search.  If my general approach seems good, then I'd like to implement a "reverse geo-coding" step that takes the coordinates from the geolocation API and uses them to get an initial value for geo_address from Google.  Using the "Display Options" popover, the user can then change that geo_address value to something more descriptive, if they'd prefer.

Other potential areas for discussion:
- I'm not positive that a Popover is the ideal UI component for this use case.  I chose it because I didn't want to over-weight the location section of the sidebar by cluttering it with options that are likely to be used fairly rarely.  I'm open to suggestions of other components, if I've just missed a better alternative.

- I added a geoIsSharedPublicly() method to PostMetadata along with a few tests.  At this point, PostMetadata has 3 geo-related methods, so it's possible that refactoring those into a separate utility would make sense at this point.

Thanks in advance for taking the time to look this over!

jsnajdr commented on client/post-editor/editor-location/options.jsx:
```diff
@@ -0,0 +1,128 @@
+			return;
+		}
+
+		this.setState( { renderContext } );
```
It's a bug to set state in a ref callback. Either set an instance property on the class, or use the [React.createRef](https://reactjs.org/docs/refs-and-the-dom.html#creating-refs) API.

griffbrad replied:
Thanks for the pointer.  This is addressed in 36843b0.

jsnajdr commented:
The `geo_public` flag was originally used, many years ago, to share the location only if you really want to. Then there was a wpcom patch r94911-wpcom that made the location always public. And `geo_public` became just a backward compatibility flag, expected to be always true on new posts.

If my undestanding is correct (pinging @beaucollins to confirm), then we don't need any UI with a checkbox: just set/unset `geo_public` whenever geolocation is added or removed and that's all.

We also don't need to add the "My Location" and "My Location is public" fields to Calypso user settings UI. They are technically a "Calypso gap", but we don't want to add them, do we?

Regarding the UI design, let's ask @folletto and @shaunandrews -- they always have awesome ideas. Could you @griffbrad post some screenshots that we can discuss?

If the "public" checkbox is removed, the popover would display only the "Description" field. And then it's better to get rid of the popover and display it inline, isn't it? Is "Description" the right name for the field, or should it be "Address"? And how does it interact with the "Search" field? Could both of them possibly be merged into one control?

I'd love to test the wp-admin "Add Location" UI but it's completely broken because of the missing `geo_location_api_key` unfortunately :cry:

There's one special case that this PR should also cover: when duplicating a post (either select "Duplicate" in the posts list or "Select a post to copy" in post editor's "More Options"), we copy the geo meta attributes to the new post. There's a `metadataWhitelist` array in `startEditingPostCopy`. Can you add the new `geo_*` attributes there?

griffbrad commented:
Thanks so much for taking the time to review and share your feedback!  I've added geo_address and geo_public to the metadataWhitelist during post copy.  I'll hold off on other tweaks and fixes until we get feedback on the necessary options and UI.  To help with the UI review, I've included a few screenshots here.

<img width="1267" alt="no-location-set" src="https://user-images.githubusercontent.com/449804/40240616-910b1838-5a7e-11e8-8fb3-541a5313ae03.png">

<img width="1261" alt="location-set-display-options-button-present" src="https://user-images.githubusercontent.com/449804/40240620-94d2fc74-5a7e-11e8-8045-b4652e728668.png">

<img width="1263" alt="location-set-popover-open" src="https://user-images.githubusercontent.com/449804/40240626-987b464c-5a7e-11e8-8f89-6057e8cb6d96.png">

griffbrad commented:
I went ahead and implemented the reverseGeocode() method we needed for getting a friendly label/description for the location after clicking "Get Current Location".  I figured that we'd need that API in place regardless of the direction we ultimately take with the UI.  I may take some time in the next day or two to fiddle with a couple other possible UI variations.

folletto commented:
Would be possible to have the "Search" field to match 1:1 what's going to be displayed, or is that going to create confusion between the data used for search and for display?

Note: this would also mean that someone couldn't type something "different" from the correct address, and I think it's a good limitation. I also think it's how WP Admin does it no?

If that's possible, then we can just show the checkbox below when the field is populated.

griffbrad commented:
@folletto Thanks for sending in some feedback!  I think that makes a lot of sense.  I've implemented that UI in c3542a0.  

<img width="1265" alt="screen shot 2018-05-22 at 3 43 34 pm" src="https://user-images.githubusercontent.com/449804/40389179-4469a880-5dd7-11e8-81b8-eef240450a9c.png">

@jsnajdr The "Display location publicly" checkbox is still included currently.  I think it makes sense to drop it at this point.  I was originally concerned because the existing posts saved with Calypso currently have "private" geo-tagging data.  However, upon further reflection, they'll remain that way even after this update unless the user explicitly returns to edit one of those existing posts and modifies the geo-location data.  

To be truly circumspect about this, I propose the following: if the user returns to one of those "private" posts, I can detect that by noting that "geo_latitude" and "geo_longitude" are set but "geo_public" is undefined.  If they modify the geo-location data on the post, I can display a message in the UI at that time explaining that the location will be displayed publicly.  That way the author has been notified before they hit "Update".  There are no surprises for anyone with that approach, and we don't clutter the UI at all for folks not impacted by this quirk.  

What do you think?

folletto commented:
> if the user returns to one of those "private" posts, I can detect that by noting that "geo_latitude" and "geo_longitude" are set but "geo_public" is undefined. If they modify the geo-location data on the post, I can display a message in the UI at that time explaining that the location will be displayed publicly. 

I agree. This is the approach that makes more sense to me. I was about to question the flag existence as well :D

griffbrad commented:
Right on.  I've removed the checkbox altogether at this point and added the little warning message when a user is editing their geo-location that is currently "private" due to the gap in older versions of Calypso.

<img width="1260" alt="screen shot 2018-05-23 at 1 09 58 pm" src="https://user-images.githubusercontent.com/449804/40443099-dae02920-5e8a-11e8-8761-d39897323439.png">

griffbrad commented:
Thanks again @jsnajdr and @folletto for the feedback so far!  I feel like the UX is in a much nicer place at this point and we've come to a good spot regarding the private/public nuances.

folletto commented:
I would suggest to not use that warning UI. I'd go for a small text right under the address when that change happens. It's not a big deal, and that control isn't very suited for that sidebar.

![screen shot 2018-05-23 at 19 21 16](https://user-images.githubusercontent.com/4389/40443517-7c65a30a-5ebe-11e8-9ba3-13ba27e66d23.png)

griffbrad commented:
That's totally fine with me.  I was mostly using the "Notice" component because it was being use when a Promise ended up in an error state.  I can certainly see how this particular string merits more subtlety, though.

<img width="264" alt="screen shot 2018-05-23 at 1 36 10 pm" src="https://user-images.githubusercontent.com/449804/40444273-80308d90-5e8e-11e8-9d6a-f200567bec4e.png">

jsnajdr commented on client/lib/post-metadata/index.js:
```diff
@@ -136,6 +136,41 @@ PostMetadata = {
+	 */
+	geoIsSharedPublicly: function( post ) {
+		if ( ! post ) {
+			return 'public';
```
What if we returned a boolean true or false instead of `public` and `private`? That's what function that ask an `is` question usually return.

griffbrad replied:
Fair enough.  I think early on I wasn't sure if I was going to end up with additional states other than just public/private, but with just those two a straight-up boolean does simplify things.  
 This is done in 07480a0 Thanks!

jsnajdr commented on client/lib/post-metadata/index.js:
```diff
@@ -136,6 +136,41 @@ PostMetadata = {
+			return 'public';
+		}
+
+		if ( 'undefined' === typeof isSharedPublicly ) {
```
You can do a simple `isSharedPublicly === undefined` comparison here. The `typeof` trick is needed only for cases where it's uncertain if the tested variable even exists, like when testing for `window` or `document`.

griffbrad replied:
Makes perfect sense.  Thanks!  This is in feb151b.

griffbrad commented on client/post-editor/editor-location/index.jsx:
```diff
@@ -100,13 +149,17 @@ class EditorLocation extends React.Component {
 			geo_latitude: toGeoString( result.geometry.location.lat ),
 			geo_longitude: toGeoString( result.geometry.location.lng ),
+			geo_address: result.formatted_address,
+			geo_public: 1,
```
Makes sense.  geo_public is assigned a string value in 762cd32.

griffbrad commented on client/post-editor/editor-location/index.jsx:
```diff
@@ -127,7 +180,17 @@ class EditorLocation extends React.Component {
+
+		if ( this.state.previouslyPrivatePostBeingModified ) {
+			publicWarning = (
+				<div class="editor-location__public-warning">
```
Fixed in 75a6f88.

griffbrad commented on client/post-editor/editor-location/style.scss:
```diff
@@ -6,6 +6,16 @@
 }
 
+.editor-location__public-warning {
+	color: #537994;
```
Added in 947bcd7.

griffbrad commented on client/post-editor/editor-location/index.jsx:
```diff
@@ -48,23 +49,71 @@ class EditorLocation extends React.Component {
+
+		reverseGeocode( latitude, longitude )
+			.then( results => {
+				const localityResults = results.filter( result => {
```
Switched to the lodash methods rather than filter/indexOf in c94f9f0.

griffbrad commented on client/post-editor/editor-location/index.jsx:
```diff
@@ -48,23 +49,71 @@ class EditorLocation extends React.Component {
+
+				recordStat( 'location_reverse_geocode_failed' );
+			} )
+			.finally( () => {
```
There are polyfills available in Babel but a second then() actually has the same effect in this case.  I switched to that approach in b012eee to avoid potential compatibility issues.  then() is maybe marginally less helpful versus finally() semantically in this case, but I think it's a reasonable trade-off.  (Here is the actual MDN compatibility table for the feature, by the way: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#Browser_compatibility)

griffbrad commented on client/lib/post-metadata/index.js:
```diff
@@ -136,6 +136,40 @@ PostMetadata = {
+	 */
+	geoIsSharedPublicly: function( post ) {
+		if ( ! post ) {
+			return true;
```
Returning null in this case in 0cf319e.

griffbrad commented on client/post-editor/editor-location/index.jsx:
```diff
@@ -48,23 +49,71 @@ class EditorLocation extends React.Component {
+		// When geo data was originally private and its coordinates have changed, we track that change
+		// so we can warn the user that after saving the information will be publicly visible.
+		if (
+			'private' === this.originalProps.isSharedPublicly &&
```
Awesome.  Thanks for pointing out that selector.  I've implemented this approach in 5d18cd4.

griffbrad commented:
Thanks so much for the detailed review @jsnajdr!  I especially appreciate the direction on the finer points of the lifecycle of state in Calypso's use of Redux and communication with the server!
